### PR TITLE
[native] Disable cache retention with NO_PREFERENCE node selection

### PR DESCRIPTION
**Pull request by zacw7**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 1]
```diff
@@ -141,6 +145,15 @@ void updateVeloxConfigs(
     configStrings.emplace(
         core::QueryConfig::kDriverCpuTimeSliceLimitMs, "1000");
   }
+
+  for (auto& entry : connectorConfigStrings) {
+    auto connectorConfig = entry.second;
+    it = connectorConfig.find("node_selection_strategy");
+    if (it == connectorConfig.end() || it->second != "SOFT_AFFINITY") {
+      connectorConfig.emplace(
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:153]: The default value for Presto java is true? Thanks!

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp [hunk 2]
```diff
@@ -141,6 +145,15 @@ void updateVeloxConfigs(
     configStrings.emplace(
         core::QueryConfig::kDriverCpuTimeSliceLimitMs, "1000");
   }
+
+  for (auto& entry : connectorConfigStrings) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/QueryContextManager.cpp:149]: Let's have a separate method updateConnectorConfigs?

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp (hunk 1 above)
Comment (zacw7) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: In Java, there are three node selection strategy:  _HARD_AFFINITY_, _SOFT_AFFINITY_ and _NO_PREFERENCE_. _NO_PREFERENCE_ is the default value.

Code pointers shared by @kewang1024 : Splits are set as `cacheable` only when the strategy is _HARD_AFFINITY_ or _SOFT_AFFINITY_.
https://github.com/prestodb/presto/blob/6e4896ddc4ed078cce1f81186e6986ecca94a141/presto-main/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java#L156-L167

And _HARD_AFFINITY_ is not used anymore, so we expect cache is enabled only when soft affinity is set and the default is without cache.

Comment (steveburnett) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: > And _HARD_AFFINITY_ is not used anymore, so we expect cache is enabled only when soft affinity is set and the default is without cache.

If _HARD_AFFINITY_ is not used anymore, should it be removed from the Presto doc in https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/develop/connectors.rst#connectorsplit ?

Comment (kewang1024) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: HARD_AFFINITY is designed for Raptor, which is not used anymore by Meta.
But open source could still use it, so we can not remove it.

Comment (tdcmeehan) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: Raptor was deleted last year by #19236

Comment (steveburnett) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: Thanks @kewang1024 and @tdcmeehan. If Raptor was removed from OSS Presto over a year ago, is there any other use of _HARD_AFFINITY_ in current Presto that it should stay in the current doc?

**Comment by kewang1024** [comment]
image
It might not be ideal, but it would be least destructive to the current workflow

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp (hunk 1 above)
Comment (kewang1024) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: I found other `HARD_AFFINITY` usage for example `LocalFileSplit`

Comment (steveburnett) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: > I found other `HARD_AFFINITY` usage for example `LocalFileSplit`

Thanks @kewang1024! In that case I think it should stay in the current doc.

**Comment by zacw7** [comment]
> One issue I can think of is if the config "node-selection-strategy" is set to "SOFT_AFFINITY" and no session property. Then on prestissimo side, it would be translated to no cache retention?

You are right, I'm surprised that Velox doesn't consider Presto system config properly. This is tricky.

File: presto-native-execution/presto_cpp/main/QueryContextManager.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply presto-native-execution/presto_cpp/main/QueryContextManager.cpp:157]: @zacw7 then change to it->second == "NO_PREFERENCE"? Also make sure it doesn't break existing meta native use case with SSD cache enabled? thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@zacw7 LGTM % comment. Thanks!

**Review by xiaoxmeng — APPROVED** [review]

**Pull request merged**